Enno focus metavariables (#714)
returntocorp/semgrep-docs: docs/experiments/multiple-focus-metavariables.md

---
slug: multiple-focus-metavariables
append_help_link: true
description: "With this rule, Semgrep matches all pieces of code captured by the focus metavariables."
---

# Using multiple focus metavariables

> Use multiple focus metavariables in one Semgrep rule by specifying each under a new `focus-metavariable` item in a YAML list. With this rule, Semgrep matches all pieces of code captured by the focus metavariables.

Author: Use multiple focus metavariables in one ~~Semgrep rule~~ by specifying each under a new `focus-metavariable` item in a YAML list

Reviewer: ```suggestion
With this syntax, Semgrep matches all pieces of code captured by the focus metavariables. Specify each metavariable that you want to focus by including it in a YAML list.
```

> :::info
> This feature is using `focus-metavariable`, see [`focus-metavariable`](/writing-rules/rule-syntax/#focus-metavariable) documentation for more information.  
> :::
> 
> <iframe src="https://semgrep.dev/embed/editor?snippet=493E" border="0" frameBorder="0" width="100%" height="432"></iframe>

Reviewer: ```suggestion
```

> There are two methods of using multiple metavariables. Each of them can lead to different results:

Reviewer: ```suggestion
There are two methods of focusing on multiple metavariables. Each of them can lead to different results:
```

- **Recommended**: Including `focus-metavariable` with more metavariables directly under the `patterns` key returns results regardless of whether there is an overlap between these variables:
    ```yaml
        patterns:
          - pattern: foo($X, ..., $Y)
          - focus-metavariable: 
            - $X
            - $Y
    ```
    See the following example:

>     TODO ADD IFRAME

Reviewer: @enncoded moved the article to experiments. Once this feature is out, let's add the Playground iframe here. :-)

Reviewer: ```suggestion
TODO - substitute placeholder code example below with simple showcase of YAML list syntax:
<iframe src="https://semgrep.dev/embed/editor?snippet=493E" border="0" frameBorder="0" width="100%" height="432"></iframe>
```

Reviewer: ```suggestion
    <iframe src="https://semgrep.dev/embed/editor?snippet=D602" border="0" frameBorder="0" width="100%" height="432"></iframe>
```

- **Advanced**: Including more `focus-metavariable` keys with different metavariables under the `pattern` key lets Semgrep find **only** the overlapping region of all the focused code:
    ```yaml
        patterns:
          - pattern: foo($X, ..., $Y)
          - focus-metavariable: $X
          - focus-metavariable: $Y
    ```
    See the following example:

>     <iframe src="https://semgrep.dev/embed/editor?snippet=e8OL" border="0" frameBorder="0" width="100%" height="432"></iframe>

Reviewer: ```suggestion
    <iframe src="https://semgrep.dev/embed/editor?snippet=W4xy" border="0" frameBorder="0" width="100%" height="432"></iframe>
```

> 

Reviewer: ```suggestion

:::tip
TODO - Among many use cases, the above **recommended** syntax allows you to simplify taint analysis rule writing. For example see the following rule:
TODO substitute - <iframe src="https://semgrep.dev/embed/editor?snippet=493E" border="0" frameBorder="0" width="100%" height="432"></iframe>
:::

```
